Cache controllers

jbartosik opened the pull request:
Cache results of attempting to determine the parent a controller.

This is to improve VPAs performance in big clusters, where we need to throttle queries we're making.

How caching is supposed to work:

* When we first try to determine parent of a controller try to get it (and cache the result).
  * This should happen rarely (when new controllers appear and when VPA starts). So it shouldn't affect VPAs performance heavily
  * It let's us avoid incorrect state
* On consequent attempts to get parent of a controller return the cached result
  * We do this frequently, so this should significantly improve our performance
* Refresh content of cache and remove unused entries in parallel

jbartosik commented:
/hold
I think it's ready for review but want to do some more tests before merging

mwielgus reviewed (changes requested):
The PR contains non trivial amount of code, but yet it doesn't have any description. Could you please provide some information what it does and why is it needed?

jbartosik commented:
> The PR contains non trivial amount of code, but yet it doesn't have any description. Could you please provide some information what it does and why is it needed?

I added a description

jbartosik commented:
/hold cancel

I ran `full-vpa` and `recommender` tests with this PR and they passed.

bskiba commented:
Looking at this now.

bskiba commented:
OK, I had a first look and looks good, thanks for splitting it into nicely separated commits, makes for a way easier review.
I'll need to take another look on Monday for a deeper review. So far I'd like to confirm some basic things.

1. When testing with 200 controllers, did you observe api call throttling?
2. Did you watch the memory consumption? Would be good to confirm we're not leaking, i.e. once we delete the controllers we should see a drop in memory after ttl passes.
3. With this solution, don't we still issue scale subresource calls when checking if the controller is scalable? https://github.com/kubernetes/autoscaler/blob/8711a1e37aa32eeda5ba7f3abb02295ab6c223db/vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_fetcher.go#L262

jbartosik commented:
> OK, I had a first look and looks good, thanks for splitting it into nicely separated commits, makes for a way easier review.
> I'll need to take another look on Monday for a deeper review. So far I'd like to confirm some basic things.
> 
> 1. When testing with 200 controllers, did you observe api call throttling?

No, I don't see any logs about throttling.

> 2. Did you watch the memory consumption? Would be good to confirm we're not leaking, i.e. once we delete the controllers we should see a drop in memory after ttl passes.

I didn't see significant change in memory usage (which is not surprising, we're storing only a couple of controller keys per controller). I'll run another test where I create a bunch of controllers to see memory usage increase, then wait to see if it drops after time.

> 3. With this solution, don't we still issue scale subresource calls when checking if the controller is scalable? https://github.com/kubernetes/autoscaler/blob/8711a1e37aa32eeda5ba7f3abb02295ab6c223db/vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_fetcher.go#L307

We do. I'll add caching for that too. Is it ok if I do this in a separate PR?

jbartosik commented:
> > 1. Did you watch the memory consumption? Would be good to confirm we're not leaking, i.e. once we delete the controllers we should see a drop in memory after ttl passes.
> 
> I didn't see significant change in memory usage (which is not surprising, we're storing only a couple of controller keys per controller). I'll run another test where I create a bunch of controllers to see memory usage increase, then wait to see if it drops after time.

I started a a few hundreds of pods. As a result memory usage increased from `23Mi` to t o`31Mi`. When I deleted pods memory usage decreased to `26Mi` and later (not sure after how much time) down to `25Mi`.  I'll repeat and see what happens

jbartosik commented:
@kgolab  PTAL

jbartosik commented:
> > 1. With this solution, don't we still issue scale subresource calls when checking if the controller is scalable? https://github.com/kubernetes/autoscaler/blob/8711a1e37aa32eeda5ba7f3abb02295ab6c223db/vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_fetcher.go#L307
> 
> We do. I'll add caching for that too. Is it ok if I do this in a separate PR?

Done

kgolab commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,149 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var now = time.Now
```
Could you please add a comment that this is done to allow controlling time from unit tests? Or is this obvious for anyone coding in golang?

jbartosik replied:
Done.

kgolab commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,149 @@
+// Usage:
+// - `Get` cached response. If there is one use it, otherwise make query and
+// - `Insert` the response you got into the cache.
+// Start two go routines:
```
Please make it clear who should start these go routines - it's not immediately clear that the caller has to do it (and I actually find it slightly strange that it's organised this way).

jbartosik replied:
Done.

kgolab commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,149 @@
+	defer cc.mux.Unlock()
+	now := now()
+	for k, v := range cc.cache {
+		if now.After(v.refreshAfter) {
```
I'd like to understand the idea behind refreshing & clearing stale entries a little better.

IIUC when an entry becomes idle, we'd refresh it few times before deleting which seems counterintuitive.
If this is true, shouldn't we throttle refreshes only to entries which were read recently? Then Get should probably return false to force a refresh upon reading an idle (& not refreshed) entry.

jbartosik replied:
I chose the simplest approach I thought would work (periodically refresh all entries, remove entries that weren't read in a while).

> If this is true, shouldn't we throttle refreshes only to entries which were read recently? Then Get should probably return false to force a refresh upon reading an idle (& not refreshed) entry.

If `Get` returns false then the entry is effectively removed. So if I understand your idea correctly it's effectively the same as setting shorter lifetime for cache entries (less than 2 refresh durations?)

kgolab replied:
If it's about simplicity then why not just:
* Insert sets validity time (let's say it's current scaleCacheEntryFreshnessTime),
* Get checks validity time and returns false if the entry is expired; the caller has to call Insert then,
* there is no background refresh,
* there is only background garbage collection?

This reduces number of calls to scaleNamespacer.Get as we never call it for entires that are no longer present.
I understand that the drawback of this solution is that all Gets are executed in the main VPA loop instead of being spread evenly over time.

So basically the discussion boils down to a question: "how long do we actively use an entry, on average"; if it's much longer than scaleCacheEntryLifetime, the amortised cost of multiple refreshes of a no-longer-used entry is low and then the gain from spreading scaleNamespacer.Get calls is likely more important.

If you think we'd indeed have long-lived entries (seems plausible) let's leave the solution as it is now.
I'd love to see some metric (or logs with counter) added at some point so we can assess the impact of this change and maybe fine-tune scaleCacheEntryLifetime vs scaleCacheEntryFreshnessTime, likely lowering the former.

kgolab commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,149 @@
+	now := now()
+	cc.cache[key] = scaleCacheEntry{
+		refreshAfter: now.Add(cc.betweenRefreshes).Add(jitter),
+		deleteAfter:  old.deleteAfter,
```
IIUC this is critical for an entry to get removed, could you please comment why it's here so nobody 'fixes' it?
I know the function comment says it's done on purpose but adding 'WHY' would help.

jbartosik replied:
Done.

kgolab commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,149 @@
+	if _, ok := cc.cache[key]; ok {
+		return
+	}
+	jitter := time.Duration(rand.Float64()*float64(cc.refreshJitter.Nanoseconds())) * time.Nanosecond
```
How about extracting this to a function?

jbartosik replied:
Done

bskiba replied:
We can also use the Jitter function from here: https://github.com/kubernetes/apimachinery/blob/master/pkg/util/wait/wait.go#L193:6

kgolab commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage_test.go:
```diff
@@ -0,0 +1,217 @@
+	assert.Equal(t, getScale(), val)
+	assert.Nil(t, err)
+
+	// Doesn't overwrite previous values
```
I'm not sure why we want this behaviour.
IIUC the result from second Insert is fresher so it should overwrite the old value (maybe just not deleteAfter).
If there is a valid reason, please add a comment about it.

jbartosik replied:
I don't expect this to come up in how we use the cache so I think this won't affect VPAs performance. And I find this easier to understand.

Please let me know if I should change the behavior

kgolab replied:
I don't have very strong feelings regarding the behaviour itself but if we simply don't care about whether it's one way of the other, maybe let's not have a test for this to basically keep it undefined?

kgolab commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_fetcher.go:
```diff
@@ -82,10 +83,41 @@ type controllerFetcher struct {
+	controllerCache controllerCacheStorage
+}
+
+func (f *controllerFetcher) periodicallyRemoveExpired(ctx context.Context, period time.Duration) {
```
Here and below - since these functions are part of controllerFetcher and not controllerCache, could you please rename them so they include 'cache' or 'controllerCache' in their names?

jbartosik replied:
Done

kgolab commented on vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:
```diff
@@ -49,7 +49,14 @@ import (
 
-const defaultResyncPeriod time.Duration = 10 * time.Minute
+const (
+	betweenScaleSubresourceRefreshes time.Duration = 10 * time.Minute
```
Maybe you could consider running this more frequently so the load (GET calls to API server) is spread more evenly in time? If GetKeysToRefresh is lightweight (and it seems so), it might be beneficial to run the whole refresher subroutine more often than main loop of the recommender.

kgolab replied:
Actually this should belong below.

jbartosik replied:
Done.

kgolab commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_fetcher.go:
```diff
@@ -82,10 +83,41 @@ type controllerFetcher struct {
+		case <-ctx.Done():
+			return
+		case <-time.After(period):
+			for _, item := range f.controllerCache.GetKeysToRefresh() {
```
Could you please add some logs here marking start & finish of this procedure? Ideally with the number of entries refreshed after the loop.

jbartosik replied:
Done.

kgolab commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_fetcher.go:
```diff
@@ -82,10 +83,41 @@ type controllerFetcher struct {
+		case <-ctx.Done():
+			return
+		case <-time.After(period):
+			f.controllerCache.RemoveExpired()
```
Maybe it would be useful here to log start & finish (or inside RemoveExpired, to get the number of entries removed).

jbartosik replied:
Done.

kgolab commented on vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:
```diff
@@ -49,7 +49,14 @@ import (
+const (
+	betweenScaleSubresourceRefreshes time.Duration = 10 * time.Minute
+	containerCacheCleanupLoopPeriod  time.Duration = time.Hour
+	containerCacheRefreshLoopPeriod  time.Duration = time.Minute
```
Seeing how these two are used I'd say they need renaming.

And seeing that I got some of these consts confused, maybe they could be commented to explain their meaning.

jbartosik replied:
Done.

kgolab commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,149 @@
+type controllerCacheStorage struct {
+	cache            map[scaleCacheKey]scaleCacheEntry
+	mux              sync.Mutex
+	betweenRefreshes time.Duration
```
I'd say it's rather validityTime not to confuse it with period between executing refresh loop but please feel free to ignore.

jbartosik replied:
Done.

kgolab commented:
> @kgolab PTAL

Many questions and some nits but I guess it's worth discussing how exactly the case is expected to behave with regards to refreshes and removal of stale entries.

jbartosik commented:
> > @kgolab PTAL
> 
> Many questions and some nits but I guess it's worth discussing how exactly the case is expected to behave with regards to refreshes and removal of stale entries.

I pushed changes with comments I applied. I had questions about a few and left them open. I'll add explanation why I want the cache to work this way and a TODO we discussed later today.

jbartosik commented:
> > > @kgolab PTAL
> > 
> > 
> > Many questions and some nits but I guess it's worth discussing how exactly the case is expected to behave with regards to refreshes and removal of stale entries.
> 
> I pushed changes with comments I applied. I had questions about a few and left them open. I'll add explanation why I want the cache to work this way and a TODO we discussed later today.

@kgolab I added the explanation (above `controllerCacheStorage`) and TODO.  Please take a look

jbartosik commented:
@kgolab @bskiba Please take a look

bskiba commented:
I've set aside some time tomorrow to take a look 👀

bskiba commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,169 @@
+	cache         map[scaleCacheKey]scaleCacheEntry
+	mux           sync.Mutex
+	validityTime  time.Duration
+	refreshJitter time.Duration
```
Might consider jitterFactor instead of jitter expressed as duration, similarly to Kuberentes wait functionality:
https://github.com/kubernetes/apimachinery/blob/master/pkg/util/wait/wait.go#L132

bskiba commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+// If key is in the cache it updates the cached value, error and refresh time (but not time to remove).
+// If the key is missing from the cache does nothing (relevant when we asynchronously updating cache and
```
maybe: (relevant when we're concurrently updating cache and removing stale entries from it..)
Also, missing commas, see comment below.

jbartosik replied:
Done.

bskiba commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,169 @@
+	now := now()
+	// We refresh entries that are waiting to be removed. So when we refresh an
+	// entry we mustn't change entries deleteAfter time (otherwise we risk never
+	// removing entries that none is reading)
```
"that noone/nobody is reading" or "that are not being read"

jbartosik replied:
Done.

bskiba commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,169 @@
+	if !ok {
+		return
+	}
+	now := now()
```
nit: we can simply use now() in line 106

jbartosik replied:
Done.

bskiba commented on vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:
```diff
@@ -49,7 +49,13 @@ import (
 
-const defaultResyncPeriod time.Duration = 10 * time.Minute
+const (
+	scaleCacheLoopPeriod time.Duration = time.Hour
```
I would like to differentiate between constants affecting whole cache (like the loop period) and cache entries (like refresh)

jbartosik replied:
Done.

bskiba commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_fetcher.go:
```diff
@@ -82,10 +83,44 @@ type controllerFetcher struct {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(period):
```
Our refresh period for an entry is 10 minutes, but we check which entries should be refreshed every 1 hour. Is this intentional?

jbartosik replied:
Changed to 1 minute

bskiba commented on vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:
```diff
@@ -49,7 +49,13 @@ import (
-const defaultResyncPeriod time.Duration = 10 * time.Minute
+const (
+	scaleCacheLoopPeriod time.Duration = time.Hour
+	scaleCacheCleanup    time.Duration = time.Hour
```
Would be nice to decide on one of lifetime and cleanup (both are used to describe the cache entry ttl AFAIU)

jbartosik replied:
Done.

bskiba commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_fetcher.go:
```diff
@@ -82,10 +83,44 @@ type controllerFetcher struct {
+	}
+}
+
+func (f *controllerFetcher) Start(ctx context.Context, removePeriod, refreshPeriod time.Duration) {
```
Could we actually for simplicity have one cache refresh function that would delete items that should be deleted and refresh the ones that are stale but still valid?

jbartosik replied:
Done.

jbartosik commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,163 @@
+// - One for refereshing cache entries, which calls `GetKeysToRefresh` then for
+//   each key makes query to the API server and calls `Refresh` to update
+//  content of the cache.
+// - Second from removing stale entries which periodically calls `RemoveExpired`
```
Done.

jbartosik commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,163 @@
+// handle such situation lifeTime should be longer than refreshAfter so the main
+// VPA loop can do its work quickly, using the cached information (instead of
+// getting stuck on refreshing the cache).
+// TODO(jbartosik): Add a way to detect when we don't refresh cache frequently
```
I was thinking about a metric reporting how much time elapsed between entries being ready for refresh and actually being refreshed. I think it would be better to implement it in a separate PR to make reviews more focused.

bskiba replied:
ack

jbartosik commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,163 @@
+
+// Returns bool indicating whether the entry was present in the cache and the cached response.
+// Updates deleteAfter for the element.
+func (cc *controllerCacheStorage) Get(namespace string, groupResource schema.GroupResource, name string) (ok bool, controller *autoscalingapi.Scale, err error) {
```
I prefer this signature because `ok` is the most important of the return values (it tells you if it makes sense to use the other two)

bskiba replied:
ack

jbartosik commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,169 @@
+	}
+}
+
+// If the key is missing from the cache updates the cached value, error and refresh time (but not time to remove).
```
Done.

jbartosik commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,169 @@
+	}
+}
+
+// If the key is missing from the cache updates the cached value, error and refresh time (but not time to remove).
```
Done.

jbartosik commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,169 @@
+
+// Removes entries from the cache which we didn't read in a while.
+func (cc *controllerCacheStorage) RemoveExpired() {
+	klog.Infof("Removing entries from controllerCacheStorage")
```
Done. We'll log it ~1/minute.

jbartosik commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,169 @@
+	}
+}
+
+// Removes entries from the cache which we didn't read in a while.
```
Done.

jbartosik commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,169 @@
+			delete(cc.cache, k)
+		}
+	}
+	klog.Infof("Removed %d entries from controllerCacheStorage", removed)
```
Done.

jbartosik commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_fetcher.go:
```diff
@@ -248,6 +284,15 @@ func (f *controllerFetcher) isWellKnown(key *ControllerKeyWithAPIVersion) bool {
+	if ok, scale, err := f.controllerCache.Get(namespace, groupResource, name); ok {
+		return scale, err
+	}
+	scale, err := f.scaleNamespacer.Scales(namespace).Get(context.TODO(), groupResource, name, metav1.GetOptions{})
```
I agree that it would be nicer to use if cache took care of fetching entries. But I think it would make code more complicated. Since this is the only place I expect to use the cache I think it's overall better to keep things simpler.

I named it cache *storage* to emphasise that it's doing less than one might hope.

bskiba replied:
That makes sense, I'm not gonna be supper stubborn about it :)

kgolab replied:
Actually it might be useful to rename the field as well, so it's clear at call site it's not a full cache but rather a storage.

jbartosik commented:
@bskiba @kgolab I pushed changes, please take a look

jbartosik commented:
@bskiba I can't reply directly to the comment about using jitter function from API machinery but I applied it

bskiba commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,164 @@
+	return ok, r.resource, r.err
+}
+
+// If key is in the cache it updates the cached value, error and refresh time (but not time to remove).
```
nit: If key is in the cache, refresh updates...

bskiba commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,164 @@
+}
+
+// If key is in the cache it updates the cached value, error and refresh time (but not time to remove).
+// If the key is missing from the cache does nothing (relevant when we're
```
nit: If the key is missing from the cache, does nothing

bskiba commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,164 @@
+//   each key makes query to the API server and calls `Refresh` to update
+//  content of the cache.
+// - Second for removing stale entries which periodically calls `RemoveExpired`
+// Each entry is refreshed after duration (`validityTime` + `jitter`) passes
```
> duration (`validityTime` + `jitter`)
This is no longer valid

jbartosik replied:
updated

bskiba commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_cache_storage.go:
```diff
@@ -0,0 +1,164 @@
+	}
+	// We refresh entries that are waiting to be removed. So when we refresh an
+	// entry we mustn't change entries deleteAfter time (otherwise we risk never
+	// removing entries are not being read).
```
nit: entries that/which are not being read

bskiba commented on vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:
```diff
@@ -49,7 +49,13 @@ import (
 
-const defaultResyncPeriod time.Duration = 10 * time.Minute
+const (
+	scaleCacheLoopPeriod         time.Duration = time.Minute
```
If we're refreshing every minute, for the usecase where there is one pod per controller, won't we have the same scalability as we have now ?
i.e. currently, if you have 600 pods, one custom controller for each, in the current solution this eats up 600 qpm = 10 qps
with the caching configured like in this PR we still take up  600 qpm = 10 qps

jbartosik replied:
Cache entry lives for `scaleCacheEntryFreshnessTime` + jitter (15 minutes on average). Having short `scaleCacheLoopPeriod` is how often we check which entries need a refresh (so it should be shorter than `scaleCacheEntryFreshnessTime` or we'll just refresh everything every loop).

bskiba replied:
Ah of course you're right, sorry

bskiba reviewed (commented):
Some nits left.

Overall I think we are close to what we want, would like to make sure the cache constants we choose work in our favor.

@kgolab Would you be able to take another look today? I would like you to explicitly approve as well once you're happy with this PR

jbartosik commented:
@bskiba  @kgolab PTAL ≤

bskiba reviewed (approved):
I'm happy, I'll LGTM once kgolab takes one more look (I think more eyes are better here as caching is always tricky :) )

bskiba commented:
And thanks a lot for the work that went into this!

jbartosik commented:
@kgolab please take a look

kgolab commented on vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:
```diff
@@ -49,7 +49,13 @@ import (
 
-const defaultResyncPeriod time.Duration = 10 * time.Minute
+const (
+	scaleCacheLoopPeriod         time.Duration = time.Minute
```
If you really want this to be spread independently from the VPA main loop, maybe choose some other period, preferably relatively prime to main loop's 60s period?
IIUC it might be a relatively short one, e.g. 13s, as the refresh loop is pretty lightweight except for scaleNamespacer.Get calls which won't change in number but instead get spread more evenly.

jbartosik replied:
Done. 7s because it's  the shortest that's in full seconds and relatively prime

kgolab commented on vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_fetcher.go:
```diff
@@ -82,10 +83,33 @@ type controllerFetcher struct {
+		case <-ctx.Done():
+			return
+		case <-time.After(period):
+			keysToReferesh := f.controllerCache.GetKeysToRefresh()
```
nit: keysToRefresh

jbartosik replied:
Done

kgolab commented:
Thanks!

I added more comments but this already looks good to me.
Please note though that some of this comments are in older conversations so might be more tricky to find, I don't know how to make github show them all together.

And I'm really sorry for taking so much time to get back to this. Please ping me earlier next time.

kgolab commented:
@bskiba , I cannot execute the command so letting you know here: LGTM

jbartosik commented:
@kgolab replies to comments I can't find here but got email notifications for:

* we don't want to refresh entries in the main loop so it wont' get stuck so I'm keeping as is (but I'm making refresh loop execute more frequently)
* I added a TODO to add something that will let us optimize performance. I'll probably do it in two stages. First I'll add something to let us determine if we need to improve performance. If we think we need performance improvements I think it's best to use something more detailed than average lifetime (e.g. I think we might want to check for changes in young collections and error responses more quickly than in collections which were around for long time, unchanged)
* I changed the test for inserting over an old value to not care which value it keeps
* renamed `controllerCache` to `scaleSubresourceCacheStorage`

bskiba commented:
Thanks a lot Joachim
/lgtm
/approve
